### PR TITLE
Clone loaded constants if they are used for inplace update

### PR DESCRIPTION
**Pull request by antiagainst**
This is a temporary solution to address the issue where we aggreate
constants in a read-only buffer but still see inplace updates to them.
Force clones for such constants for now so that down the pipeline we
will create buffers for them.

### CODE REVIEW

**Comment by antiagainst** [comment]
Ben: I think this implements your suggestion at https://github.com/google/iree/pull/6049#pullrequestreview-673245396. Let me know otherwise.

File: iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp [hunk 1]
```diff
@@ -159,6 +176,19 @@ struct InsertImmutabilityPreservingStreamClones
         });
         didClone = true;
       }
+
+      // TODO(GH-5492): This is a temporary solution to address the issue
```
Comment (benvanik) [inline iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp:180]: nit: use TODO(#issue) so github will automatically link the issue number
```suggestion
      // TODO(#5492): This is a temporary solution to address the issue
```

File: iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp [hunk 2]
```diff
@@ -101,6 +101,23 @@ static bool hasUsersInStreamAfterUpdate(Value value, Operation *updateOp) {
   return false;
 }
 
+// Returns true if the given |operand| is a constant tied to a result of
+// |updateOp| and the |updateOp| has inplace update semantics.
+static bool updatesConstantInStream(Value operand, Operation *updateOp) {
+  // Only two ops have inplace update semantics thus far. (TensorReshapeOp,
+  // which also implements TiedOpInterface, is fine.) Checking the explicit
+  // op list is not good; we should have an op interface.
+  if (!isa<DispatchOp, TensorUpdateOp>(updateOp)) return false;
```
Comment (benvanik) [inline iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp:110]: I think we should work toward using memory effects for this - did that ever get finished in MLIR? Being able to query whether `updateOp` has a write effect on `operand` would make this much less brittle. Do you have any experience with it?
(I thought I had an issue we could tag as a TODO(#) here but I can't seem to find it - will keep looking)
```suggestion
  // Only two ops have inplace update semantics thus far. (TensorReshapeOp,
  // which also implements TiedOpInterface, is fine.) Checking the explicit
  // op list is not good; we should use memory effects to tell when the operand
  // is modified by the op.
  if (!isa<DispatchOp, TensorUpdateOp>(updateOp)) return false;
```

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp:110]: Good points about side effect! Forgot about that beast. Though I'm not sure how likely it applies here, at least for `flow.dispatch`. AFAICT, side effect annotation is static in ODS, meaning an operand either has it or not. But in our case it's more dynamic, meaning we may or may not tie an operand to an result and that can change with transformations. If not tied, there is no write side effect. Also, all the operands that can be tied in `flow.dispatch` is in one variadic definition..

Comment (antiagainst) [reply iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp:110]: Actually, I'm not even sure what a write side effect to a tensor means. Tensor by definition is one read-only integrated component. Those inplace "updates" are more a "hint" that the underlying buffer can be shared in my understanding.

**Review by benvanik — APPROVED** [review]

**Comment by GMNGeoffrey** [comment]
Thanks! I confirmed this fixes the issue I was hitting

**Pull request merged**